crates/leon/src/parser.rs

```diff
@@ -0,0 +1,588 @@
+use std::{borrow::Cow, mem::replace};
+
+use crate::{Item, Literal, ParseError, Template};
+
+#[derive(Debug, Clone, Copy)]
+enum Token {
+    Text {
+        start: usize,
+        end: usize,
+    },
+    BracePair {
+        start: usize,
+        key_seen: bool,
+        end: usize,
+    },
+    Escape {
+        start: usize,
+        end: usize,
+        ch: Option<char>,
+    },
+}
+
+impl Token {
+    fn start_text(pos: usize, ch: char) -> Self {
+        Self::Text {
+            start: pos,
+            end: pos + ch.len_utf8(),
+        }
+    }
+
+    fn start_text_single(pos: usize) -> Self {
+        Self::Text {
+            start: pos,
+            end: pos,
+        }
+    }
+
+    fn start_brace_pair(pos: usize, ch: char) -> Self {
+        Self::BracePair {
+            start: pos,
+            key_seen: false,
+            end: pos + ch.len_utf8(),
+        }
+    }
+
+    fn start_escape(pos: usize, ch: char) -> Self {
+        Self::Escape {
+            start: pos,
+            end: pos + ch.len_utf8(),
+            ch: None,
+        }
+    }
+
+    fn is_empty(&self, source_len: usize) -> bool {
+        match self {
+            Self::Text { start, end } => {
+                *start >= source_len || *end >= source_len || *start > *end
+            }
+            Self::BracePair {
+                start,
+                key_seen,
+                end,
+            } => !key_seen || *start >= source_len || *end >= source_len || *start > *end,
+            Self::Escape { start, end, .. } => {
+                *start >= source_len || *end >= source_len || *start > *end
+            }
+        }
+    }
+
+    fn start(&self) -> usize {
+        match self {
+            Self::Text { start, .. }
+            | Self::BracePair { start, .. }
+            | Self::Escape { start, .. } => *start,
+        }
+    }
+
+    fn end(&self) -> usize {
+        match self {
+            Self::Text { end, .. } | Self::BracePair { end, .. } | Self::Escape { end, .. } => *end,
+        }
+    }
+
+    fn set_end(&mut self, pos: usize) {
+        match self {
+            Self::Text { end, .. } | Self::BracePair { end, .. } | Self::Escape { end, .. } => {
+                *end = pos
+            }
+        }
+    }
+
+    fn debug<'a>(&'a self, source: &'a str) -> (&str, &Self) {
+        (
+            if self.is_empty(source.len()) {
+                ""
+            } else {
+                &source[(self.start())..=(self.end())]
+            },
+            self,
+        )
+    }
+}
+
+impl<'s> Template<'s> {
+    #[allow(clippy::should_implement_trait)] // TODO: implement FromStr
+    pub fn from_str(s: &'s str) -> Result<Self, ParseError<'s>> {
+        Self::parse_items(s).map(|items| Template {
+            items: Cow::Owned(items),
+            default: None,
+        })
+    }
+
+    fn parse_items(s: &'s str) -> Result<Vec<Item<'s>>, ParseError<'s>> {
+        let source_len = s.len();
```

Comment:
I think using [`nom`](https://docs.rs/nom) here would significantly simplify the code.

Reply:
I've gotten this far without dependencies for the parser... !

Reply:
I still wish that we could have simpler code here, whether it's using nom or other methods.